fix(lsp): take splits into account for `vim.lsp.util.make_floating_popup_opts`

dhruvmanila opened the pull request:
`vim.lsp.util.make_floating_popup_options` would give the incorrect anchor as splits were not taken into account in the calculation. This will make sure the exact cursor position w.r.t to the editor is taken.

This is a single vertical split where the cursor is on the right window and in the previous version, the anchor would be in the west direction even though that would result in shifting the window as it would reach the right edge.

https://user-images.githubusercontent.com/67177269/118445376-fc478180-b70b-11eb-96ea-0659dfed0ced.mov

New version will make sure to take splits into account and get the correct anchor:

https://user-images.githubusercontent.com/67177269/118445394-010c3580-b70c-11eb-9131-75fa27bd7c1c.mov

dhruvmanila committed:
fix(lsp): take splits into account for floating opts

`vim.lsp.util.make_floating_popup_options` would give the incorrect
anchor as splits were not taken into account in the calculation. This
will make sure the exact cursor position w.r.t to the editor is taken.

weilbith commented:
I would recommend to make it even more precise if already working on it. You can use the API function `nvim_list_uis()` to get the absolute dimensions. This will help to improve also the calculation for the lines below and columns after.

dhruvmanila committed:
fix(lsp): use total lines of the editor

To calculate the number of lines below, we should subtract the number of
lines above from the total lines of the editor instead of the current
window.

dhruvmanila commented:
@weilbith `nvim_list_uis` gives the absolute dimension of the Neovim window and I don't see the difference between using that and `vim.o.lines` and `vim.o.columns` to get the same information.

In the implementation, only `vim.o.columns` is needed, as we're only interested in checking if the total number of columns up 
 to the cursor + width of the popup window surpasses the available columns (`vim.o.columns`).

For lines, I see the calculation is wrong (thanks for the suggestion), so instead of `vim.fn.winheight(0)` which is the height of the current window (not taking into account if there are any horizontal splits), we will use `vim.o.lines` which is same as what we would be getting from `nvim_list_uis`

weilbith commented:
What is blocking this PR?

dhruvmanila commented:
I don't really know.

@mjlbach Hi, if possible, can you take a look at this PR? It's a small change and I've documented the changes in the commit messages as well.

Thanks if you do :)

mjlbach commented:
What is the intended behavior? I have not looked into this at all, but the behavior of this PR does not seem correct.

<img width="1279" alt="image" src="https://user-images.githubusercontent.com/13316262/125110601-a226c500-e099-11eb-90c1-c0f10b9063d8.png">

dhruvmanila commented:
Thanks for taking a look at it. This PR only fixes which corner the floating window will anchor to. The issue regarding floating window on top of the cursor was present even before this PR (I will take a look at it too) as you can see in the _before_ video.

_Before:_
* The diagnostics window is anchored to North while there is more space available on the top taking into account the height of the floating window as well. This will be more easily visible in hover popup where the height is more.
* The hover popup is also anchored to North even though the window does not fit as you can see the window is pushed upwards. (https://github.com/neovim/neovim/pull/14569/commits/59d9068a31261fb1ead7d200b0e3b06920c06407 fixes that by using the editor lines instead of window lines)
* As I decrease the width of the bottom right window, the space available for the hover window on the right side decreases and so it should anchor to east but it does not.

<details>
<summary><b>Before video</b></summary>

https://user-images.githubusercontent.com/67177269/125149093-7e748500-e154-11eb-801a-5b50fdb54246.mov

</details>

_After:_
* Here you can see the diagnostics window is anchored to the South direction as expected when taking the editor lines into account instead of just window lines.
* The hover popup is correctly anchored to the South as there is more space available on the top.
* And the East/West decision is also taken correctly with using the full editor width to calculate the anchor.

<details>
<summary><b>After video</b></summary>

https://user-images.githubusercontent.com/67177269/125149096-82a0a280-e154-11eb-833b-988156fd8434.mov

</details>

I hope this clarifies what I did in this PR, if not feel free to ask any questions you might have :)

mfussenegger commented:
Sorry if this went under the radar. I think improving the anchoring is something we should still do, but in light of https://github.com/neovim/neovim/issues/25272 and https://github.com/neovim/neovim/issues/25514 I'd put these efforts into whatever is happening with vim.ui or nvim_open_win
